File: Schemes.Manager/CustomerManager.cs
using Newtonsoft.Json;
using Schemes.Models;
using Schemes.Repository;
using Schemes.ViewModels;
using System.Collections.Generic;
using System.Text;

namespace Schemes.Manager
{
    public class CustomerManager
    {
        private readonly SchemesContext _dbContext;
        public CustomerManager(SchemesContext DbContext)
        {
            _dbContext = DbContext;
        }
        public List<SchemeDetails> GetSchemesList(string name, string langCode)
        {
            List<SchemeDetails> schemeDetails = new CustomerRepository(_dbContext).GetSchemesList(name);
            return schemeDetails;
        }
        public List<SchemeDetails> GetSchemesListByLangCode(string name, string langCode)
        {
            List<SchemeDetails> schemeDetails = new CustomerRepository(_dbContext).GetSchemesListByLangCode(name, langCode);
            return schemeDetails;
        }
        public List<SchemeDetails> GetAllSchemesList()
        {
            List<SchemeDetails> schemeDetails = new CustomerRepository(_dbContext).GetAllSchemesList();
            return schemeDetails;
        }
        public bool AddNewScheme(SchemeDetails schemeDetails)
        {
            var schemeID = new CustomerRepository(_dbContext).AddNewScheme(schemeDetails);
            var list = new List<string>();
            list.Add(schemeDetails.NameOftheScheme);
            list.Add(schemeDetails.Description);
            list.Add(schemeDetails.EligibilityCriteria);
            list.Add(schemeDetails.Benefits);
            list.Add(schemeDetails.Area);
            list.Add(schemeDetails.DocumentsRequired);
            list.Add(schemeDetails.ApplyAndLink);
            list.Add(schemeDetails.AvailableFor);

            TranslationAPI(list, schemeID);
            return true;
        }
        public List<CustomerDetails> GetCustomersList()
        {
            var list = new CustomerRepository(_dbContext).GetCustomersList();
            return list;
        } 
        public SchemeDetails GetSchemesDetails(int schemeiD)
        {
            var scheme = new CustomerRepository(_dbContext).GetSchemesDetails(schemeiD);
            return scheme;
        }
        public bool UpdateScheme(SchemeDetails schemeDetails)
        {
            var result = new CustomerRepository(_dbContext).UpdateScheme(schemeDetails);
            return result;
        }
        public bool DeleteScheme(int id)
        {
            var result = new CustomerRepository(_dbContext).UpdateSchemeStatus(id);
            return result;
        }
        public bool UpdateCustomerStatus(int customerID)
        {
            var result = new CustomerRepository(_dbContext).UpdateCustomerStatus(customerID);
            return result;
        }
        public LoanDetails GetLoanDetails(string bankName)
        {
            var loanDetails = new CustomerRepository(_dbContext).GetLoanDetails(bankName);
            return loanDetails;
        }
        public List<string> GetLoantypes(string bankName)
        {
            var types = new CustomerRepository(_dbContext).GetLoantypes(bankName);
            return types;
        }
        public List<LoanDetails> GetAllBanksList()
        {
            List<LoanDetails> loanDetails = new CustomerRepository(_dbContext).GetLoanDetailsList();
            return loanDetails;
        }
        public bool AddBanksLoanDetails(LoanDetails loanDetails)
        {
            var schemeID = new CustomerRepository(_dbContext).AddLoanDetails(loanDetails);
            
            return true;
        }
        public LoanDetails? GetLoanDetailsByID(int Id)
        {
            var details = new CustomerRepository(_dbContext).GetLoanDetailsList().Where(s=>s.LoanInterestDetailID==Id).FirstOrDefault();
            return details;
        }
        public bool UpdateLoanDetails(LoanDetails loanDetails)
        {
            var result = new CustomerRepository(_dbContext).UpdateLoanDetails(loanDetails);
            return result;
        }
        public async Task TranslationAPI(List<string> textsTotranslate,int schemeID)
        {
            string endpoint = "https://api.cognitive.microsofttranslator.com";
            string subscriptionKey = "3276690b2ef842c29b6eee6cef002467";
            string region = "eastus";
            string[] targetLanguages = { "hi","kn", "ml", "ta", "te","ur" };
            var requestBody = new List<object>();

            foreach (string targetLanguage in targetLanguages)
            {
                foreach (var text in textsTotranslate)
                {
                    requestBody.Add(new { Text = text });
                }
                string url = $"{endpoint}/translate?api-version=3.0&to={targetLanguage}";

                using (HttpClient client = new HttpClient())
                {
                    using (HttpRequestMessage request = new HttpRequestMessage(HttpMethod.Post, url))
                    {
                        string requestBodyJson = Newtonsoft.Json.JsonConvert.SerializeObject(requestBody);

                        client.DefaultRequestHeaders.Add("Ocp-Apim-Subscription-Key", subscriptionKey);
                        client.DefaultRequestHeaders.Add("Ocp-Apim-Subscription-Region", region);
                        request.Content = new StringContent(requestBodyJson, Encoding.UTF8, "application/json");

                        HttpResponseMessage response = await client.SendAsync(request);
                        string responseBody = await response.Content.ReadAsStringAsync();

                        if (response.IsSuccessStatusCode)
                        {
                            var translations = JsonConvert.DeserializeObject<List<TranslationResponse>>(responseBody);
                            var schemeData = new MultilingualSchemesDataVM();

                            for (int i = 0; i < translations.Count; i++)
                            {
                                string translatedText = translations[i].translations.First().text ?? ""; // Use the first translation as translated text
                                string translatedLang = translations[i].translations.First().to ?? ""; // Use the first translation as translated text

                                switch (i)
                                {
                                    case 0:
                                        schemeData.NameOftheScheme = translatedText;
                                        break;
                                    case 1:
                                        schemeData.Description = translatedText;
                                        break;
                                    case 2:
                                        schemeData.EligibilityCriteria = translatedText;
                                        break;
                                    case 3:
                                        schemeData.Benefits = translatedText;
                                        break;
                                    case 4:
                                        schemeData.Area = translatedText;
                                        break;
                                    case 5:
                                        schemeData.DocumentsRequired = translatedText;
                                        break;
                                    case 6:
                                        schemeData.ApplyAndLink = translatedText;
                                        break;
                                    case 7:
                                        schemeData.AvaliableFor = translatedText;
                                        break;
                                    
                                    default:
                                        // Handle cases where there are more translations than properties
                                        break;
                                }
                            }
                            new CustomerRepository(_dbContext).AddNewSchemeByLangCode(schemeData, schemeID, targetLanguage);

                        }
                    }
                }
            }
        }
        public bool CreateNewAdmin(AdminRegistrationVM adminRegistrationVM)
        {
            var schemeID = new CustomerRepository(_dbContext).CreateNewAdmin(adminRegistrationVM);

            return true;
        }
    }
}

File: schemes-python-backend/app/managers/customer_manager.py
from sqlalchemy.orm import Session
from typing import List, Optional
from datetime import datetime
import requests
import asyncio
from app.repositories.customer_repository import CustomerRepository
from app.schemas import (
    SchemeDetailsCreate, SchemeDetailsUpdate, SchemeDetails,
    Customer, LoanDetailsCreate, LoanDetails,
    AdminCreate, EMICalculationRequest, EMICalculationResponse
)

class CustomerManager:
    def __init__(self, db: Session):
        self.db = db
        self.customer_repo = CustomerRepository(db)
    
    def get_schemes_list(self, name: str, lang_code: str = "en") -> List[dict]:
        """Get schemes list with optional language support"""
        try:
            if lang_code and lang_code != "en":
                return self.customer_repo.get_schemes_list_by_lang_code(name, lang_code)
            else:
                return self.customer_repo.get_schemes_list(name)
        except Exception as e:
            print(f"Get schemes list error: {e}")
            raise e
    
    def get_schemes_list_by_lang_code(self, name: str, lang_code: str) -> List[dict]:
        """Get schemes list by language code"""
        try:
            return self.customer_repo.get_schemes_list_by_lang_code(name, lang_code)
        except Exception as e:
            print(f"Get schemes list by lang code error: {e}")
            raise e
    
    def get_all_schemes_list(self) -> List[dict]:
        """Get all active schemes"""
        try:
            return self.customer_repo.get_all_schemes_list()
        except Exception as e:
            print(f"Get all schemes list error: {e}")
            raise e
    
    def add_new_scheme(self, scheme_details: SchemeDetailsCreate) -> bool:
        """Add new scheme with multilingual support"""
        try:
            scheme_id = self.customer_repo.add_new_scheme(scheme_details)
            
            # Prepare texts for translation
            texts_to_translate = [
                scheme_details.name_of_the_scheme,
                scheme_details.description,
                scheme_details.eligibility_criteria,
                scheme_details.benefits,
                scheme_details.area,
                scheme_details.documents_required,
                scheme_details.apply_and_link,
                scheme_details.available_for
            ]
            
            # Call translation API asynchronously
            asyncio.create_task(self.translation_api(texts_to_translate, scheme_id))
            
            return True
        except Exception as e:
            print(f"Add new scheme error: {e}")
            raise e
    
    def get_customers_list(self) -> List[Customer]:
        """Get all customers"""
        try:
            return self.customer_repo.get_customers_list()
        except Exception as e:
            print(f"Get customers list error: {e}")
            raise e
    
    def get_schemes_details(self, scheme_id: int) -> Optional[dict]:
        """Get specific scheme details"""
        try:
            return self.customer_repo.get_schemes_details(scheme_id)
        except Exception as e:
            print(f"Get schemes details error: {e}")
            raise e
    
    def update_scheme(self, scheme_id: int, scheme_details: SchemeDetailsUpdate) -> bool:
        """Update existing scheme"""
        try:
            return self.customer_repo.update_scheme(scheme_id, scheme_details)
        except Exception as e:
            print(f"Update scheme error: {e}")
            raise e
    
    def delete_scheme(self, scheme_id: int) -> bool:
        """Delete (deactivate) scheme"""
        try:
            return self.customer_repo.update_scheme_status(scheme_id)
        except Exception as e:
            print(f"Delete scheme error: {e}")
            raise e
    
    def update_customer_status(self, customer_id: int) -> bool:
        """Update customer status"""
        try:
            return self.customer_repo.update_customer_status(customer_id)
        except Exception as e:
            print(f"Update customer status error: {e}")
            raise e
    
    def get_loan_details(self, bank_name: str, loan_type: str) -> Optional[LoanDetails]:
        """Get loan details"""
        try:
            return self.customer_repo.get_loan_details(bank_name, loan_type)
        except Exception as e:
            print(f"Get loan details error: {e}")
            raise e
    
    def get_loan_types(self, bank_name: str) -> List[str]:
        """Get loan types for a bank"""
        try:
            return self.customer_repo.get_loan_types(bank_name)
        except Exception as e:
            print(f"Get loan types error: {e}")
            raise e
    
    def get_all_banks_list(self) -> List[LoanDetails]:
        """Get all loan details"""
        try:
            return self.customer_repo.get_loan_details_list()
        except Exception as e:
            print(f"Get all banks list error: {e}")
            raise e
    
    def add_banks_loan_details(self, loan_details: LoanDetailsCreate) -> bool:
        """Add new loan details"""
        try:
            return self.customer_repo.add_loan_details(loan_details)
        except Exception as e:
            print(f"Add banks loan details error: {e}")
            raise e
    
    def get_loan_details_by_id(self, loan_id: int) -> Optional[LoanDetails]:
        """Get loan details by ID"""
        try:
            loans = self.customer_repo.get_loan_details_list()
            for loan in loans:
                if loan.loan_interest_detail_id == loan_id:
                    return loan
            return None
        except Exception as e:
            print(f"Get loan details by ID error: {e}")
            raise e
    
    def update_loan_details(self, loan_id: int, loan_details: LoanDetailsCreate) -> bool:
        """Update loan details"""
        try:
            return self.customer_repo.update_loan_details(loan_id, loan_details)
        except Exception as e:
            print(f"Update loan details error: {e}")
            raise e
    
    def create_new_admin(self, admin_data: AdminCreate) -> bool:
        """Create new admin"""
        try:
            return self.customer_repo.create_new_admin(admin_data)
        except Exception as e:
            print(f"Create new admin error: {e}")
            raise e
    
    def calculate_emi(self, calculation_request: EMICalculationRequest) -> EMICalculationResponse:
        """Calculate EMI for loan"""
        try:
            principal = calculation_request.loan_amount
            annual_rate = calculation_request.interest_rate / 100
            monthly_rate = annual_rate / 12
            months = calculation_request.loan_tenure_months
            
            # EMI calculation formula
            if monthly_rate > 0:
                emi = principal * (monthly_rate * (1 + monthly_rate) ** months) / ((1 + monthly_rate) ** months - 1)
            else:
                emi = principal / months
            
            total_payment = emi * months
            total_interest = total_payment - principal
            
            return EMICalculationResponse(
                monthly_emi=round(emi, 2),
                total_interest=round(total_interest, 2),
                total_payment=round(total_payment, 2)
            )
        except Exception as e:
            print(f"Calculate EMI error: {e}")
            raise e
    
    async def translation_api(self, texts_to_translate: List[str], scheme_id: int):
        """Translate scheme details to multiple languages"""
        try:
            # Supported languages for translation
            target_languages = ["hi", "te", "ta", "bn", "mr", "gu"]  # Hindi, Telugu, Tamil, Bengali, Marathi, Gujarati
            
            for lang_code in target_languages:
                try:
                    # Call translation service (placeholder - integrate with Google Translate API, Azure Translator, etc.)
                    translated_texts = await self._translate_texts(texts_to_translate, lang_code)
                    
                    if translated_texts and len(translated_texts) >= 8:
                        # Create multilingual scheme record
                        from app.schemas import MultilingualSchemeCreate
                        multilingual_scheme = MultilingualSchemeCreate(
                            name_of_the_scheme=translated_texts[0],
                            description=translated_texts[1],
                            eligibility_criteria=translated_texts[2],
                            benefits=translated_texts[3],
                            area=translated_texts[4],
                            documents_required=translated_texts[5],
                            apply_and_link=translated_texts[6],
                            available_for=translated_texts[7],
                            lang_code=lang_code
                        )
                        
                        self.customer_repo.add_new_scheme_by_lang_code(
                            multilingual_scheme, scheme_id, lang_code
                        )
                        
                except Exception as lang_error:
                    print(f"Translation error for {lang_code}: {lang_error}")
                    continue
                    
        except Exception as e:
            print(f"Translation API error: {e}")
    
    async def _translate_texts(self, texts: List[str], target_language: str) -> List[str]:
        """Translate texts using external translation service"""
        # Placeholder for translation service integration
        # In production, integrate with Google Translate API, Azure Translator, etc.
        
        try:
            # Example using Google Translate API (you need to set up credentials)
            # from google.cloud import translate_v2 as translate
            # translate_client = translate.Client()
            # 
            # translated_texts = []
            # for text in texts:
            #     if text:
            #         result = translate_client.translate(text, target_language=target_language)
            #         translated_texts.append(result['translatedText'])
            #     else:
            #         translated_texts.append(text)
            # 
            # return translated_texts
            
            # For now, return original texts (placeholder)
            print(f"Translating {len(texts)} texts to {target_language}")
            return texts
            
        except Exception as e:
            print(f"Translation service error: {e}")
            return texts
